Remove VC response signing and fix HTTP error handling

eserilev opened the pull request:
## Issue Addressed

#4597 and #5423

## Proposed Changes

- Removed VC response signing and related code paths
- prevent warp backtracking issues by changing our usage of `and_then` to `then`

Closes #4597
Closes #5423

eserilev committed:
and_then to then
remove expect
move convert_rejection to utils
remove signer from vc api

eserilev committed:
remove key

eserilev committed:
remove auth header

michaelsproul commented on common/eth2/src/lighthouse_vc/http_client.rs:
```diff
@@ -78,14 +67,11 @@ pub fn parse_pubkey(secret: &str) -> Result<Option<PublicKey>, Error> {
             server,
-            server_pubkey: parse_pubkey(&secret)?,
-            secret: Some(secret.into()),
-            authorization_header: AuthorizationHeader::Bearer,
```
Ooh, this is deleting a little too much. We need to keep the `Authorization` header on both the client and server side. We just need to change the authentication token to an arbitrary byte string, rather than a token derived from the server's signing key.

michaelsproul commented on validator_client/src/http_api/api_secret.rs:
```diff
@@ -1,211 +0,0 @@
-/// For users, this public key is a "secret" that can be shared with API consumers to provide them
-/// access to the API. We avoid calling it a "public" key to users, since they should not post this
-/// value in a public forum.
-pub const PK_FILENAME: &str = "api-token.txt";
```
For backwards compatibility, lets keep loading the contents of the `api-token.txt` and using that as the token. We can just ignore the old `.secp-sk` file which contains the secret key.

To generate new tokens we can use a random string of bytes, rather than `api-token-0xpubkey`

eserilev committed:
revert

eserilev committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into vc-api-fix

eserilev committed:
merge unstable

eserilev committed:
revert

eserilev committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into vc-api-fix

eserilev committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into vc-api-fix

eserilev committed:
refactor blocking json task

eserilev committed:
linting

eserilev committed:
revert logging

eserilev commented:
I've reverted some of my deletions and refactored a few things

### Removing VC response signing

I've removed VC response signing which makes VC response error handling much simpler. Everything is now wrapped in a `blocking_json_task` which also converts warp errors into `Responses` to prevent weird warp backtrace nonsense

### VC Auth header

I've reverted the changes that removed VC auth header logic. 

> We just need to change the authentication token to an arbitrary byte string, rather than a token derived from the server's signing key.

In `ApiSecret::create_or_open` 

https://github.com/sigp/lighthouse/blob/afccdf84b1b9067ea35acc4f4cededf769574c7f/validator_client/src/http_api/api_secret.rs#L45-L47

This little snippet already randomly generates an auth token. I've removed the following logic

- Generating the server signing key from the api token
- Storing/loading the server signing key on/from disk
- Checking that the server signing key and api token are a valid key pair

I've also ensured that we're still backwards compatible, by loading the api token from disk if it already exists

eserilev committed:
remove response signing checks in validtor http_api client

michaelsproul commented on validator_client/src/http_api/api_secret.rs:
```diff
@@ -28,47 +18,34 @@ pub const PK_FILENAME: &str = "api-token.txt";
-        if !(sk_path.exists() && pk_path.exists()) {
+        if !pk_path.exists() {
             let sk = SecretKey::random(&mut thread_rng());
             let pk = PublicKey::from_secret_key(&sk);
```
I think we should go a little further and remove the concept of the token being a public key. We should just generate a random string if the file doesn't exist.

If it does exist, then we shouldn't expect it to start with `PK_PREFIX` (I think we should delete `PK_PREFIX` from the codebase) and we shouldn't try to parse it as hex or using `PublicKey::parse_compressed`.

This will be backwards-compatible because we can support any arbitrary string as the token:

```
# fine
api-token-0xabcdef...ffff
# also fine
hello_world
# also fine
567asd9768sad786sa8d7687as6d
```

It's also more flexible, because people defining their own token don't need to worry about adding the prefix (easy enough), or making the key a valid secp256k1 key (more complicated).

eserilev replied:
I've removed the concept of public tokens, pk prefix, and hex parsing. The api key should be backwards compatible with existing keys, and for new keys it should just be a randomly generated string. There was also some minor refactoring done to remove some redundant fields

changes are in f270e02

michaelsproul commented on common/eth2/src/lighthouse_vc/http_client.rs:
```diff
@@ -263,7 +230,10 @@ impl ValidatorClientHttpClient {
-            Ok(resp) => self.signed_json(resp).await.map(Option::Some),
+            Ok(resp) => {
+                let body = resp.bytes().await.map_err(Error::from)?;
+                serde_json::from_slice(&body).map_err(Error::InvalidJson)
```
I think you need the `.map(Option::Some)` here, otherwise `serde_json` will try to decode an `Option<T>` rather than a `T`. It's probably fine most of the time, but might be weird if the server sent back `null` which could decode to `None`.

eserilev replied:
fixed in f270e02

michaelsproul reviewed (changes requested):
Looking good! I think we can get this in for v5.2.0 pretty soon

eserilev committed:
remove notion of public key, prefixes, and simplify token generation

eserilev committed:
fmt

michaelsproul commented on validator_client/src/http_api/api_secret.rs:
```diff
@@ -1,164 +1,76 @@
-/// Length of the raw secret key, in bytes.
-pub const SK_LEN: usize = 32;
-
 /// The name of the file which stores the public key.
 ///
 /// For users, this public key is a "secret" that can be shared with API consumers to provide them
 /// access to the API. We avoid calling it a "public" key to users, since they should not post this
 /// value in a public forum.
```
```suggestion
/// The name of the file which stores the API token.
```

michaelsproul reviewed (approved):
Looks great, let's do it!

michaelsproul committed:
Remove outdated comment on public key

michaelsproul commented:
Marking as `backwards-incompat` because removing the `Signature` header is technically a breaking change for the ~0 users who care about this :grin:

Let's merge

michaelsproul commented:
@mergify queue
